File: ruby/lib/user_interface/user_interface.rb
require_relative './help_command'
require_relative './unknown_command'
require_relative './user_input'

module UserInterface

	class UserInterface

		def initialize(catalogue_actions, warehouse_actions, basket_actions)
			@catalogue_actions = catalogue_actions
			@warehouse_actions = warehouse_actions
			@basket_actions = basket_actions
		end

		def start
			while 1 do
				print "shop> "
				$stdout.flush
				begin
					String line = gets
					if line.nil?
						puts
						return
					end
					line = line.strip
					unless line.empty?
						return if line[0] == 'q'
						cmd = UserInput.new(line)
						createHandler(cmd.command).run(cmd)
					end
				rescue StandardError => e
					$stderr.puts "ERROR: #{e}"
				end
			end
		end

		def createHandler(command)
			case (command)
			when 'a'
				return @basket_actions.add_to_basket
			when 'b'
				return @basket_actions.display_basket
			when 'c'
				return @basket_actions.checkout
			when 'd'
				return @catalogue_actions.display_product_details
			when 'h'
				return HelpCommand.new
			when 'p'
				return @catalogue_actions.list_products
			when 'r'
				return @warehouse_actions.replenish_stock
			when 's'
				return @warehouse_actions.stock_report
			else
				return UnknownCommand.new
			end
		end

	end
end

File: python3/user_interface/user_interface.py
from . user_input import UserInput
from . help_command import HelpCommand
from . unknown_command import UnknownCommand
from . usage_exception import UsageException
from warehouse_actions.unknown_product_exception import UnknownProductException

class UserInterface:

    def __init__(self, catalogueActions, warehouseActions, basketActions):
        self.catalogueActions = catalogueActions
        self.warehouseActions = warehouseActions
        self.basketActions = basketActions

    def start(self):
        while True:
            try:
                line = input("shop> ")
                line = line.strip()
                if line:
                    if line[0] == 'q': return
                    cmd = UserInput(line)
                    self.createHandler(cmd.command).run(cmd)
            except Exception as e:
                print("ERROR: " + str(e))

    def createHandler(self, command):
        if command == 'a':
            return self.basketActions.addToBasket()
        elif command == 'b':
            return self.basketActions.displayBasket()
        elif command == 'c':
            return self.basketActions.checkout()
        elif command == 'd':
            return self.catalogueActions.displayProductDetails()
        elif command == 'h':
            return HelpCommand()
        elif command == 'p':
            return self.catalogueActions.listProducts()
        elif command == 'r':
            return self.warehouseActions.replenishStock()
        elif command == 's':
            return self.warehouseActions.stockReport()
        else:
            return UnknownCommand()

